Mathlib/Analysis/Calculus/FirstDerivativeTest.lean

```diff
@@ -0,0 +1,176 @@
+    fun _ hx ↦ (Ioo_union_left g₁ ▸ hx).elim
+    (fun hx ↦ (hd₁.differentiableAt <| Ioo_mem_nhds hx.1 hx.2).continuousAt.continuousWithinAt)
+    (fun hx ↦ mem_singleton_iff.1 hx ▸ h.continuousWithinAt)
+  isLocalMax_of_mono_anti g₀ g₁
+    (monotoneOn_of_deriv_nonneg (convex_Ioc a b)
+    continuous_Ioc (by simp_all) (by simp_all))
+    (antitoneOn_of_deriv_nonpos (convex_Ico b c)
+    continuous_Ico (by simp_all) (by simp_all))
+
+/-- The First-Derivative Test from calculus, minima version. -/
+lemma first_derivative_test_Ioo_min {f : ℝ → ℝ} {a b c : ℝ} (h : ContinuousAt f b)
+    (g₀ : a < b) (g₁ : b < c)
+    (hd₀ : DifferentiableOn ℝ f (Set.Ioo a b)) (hd₁ : DifferentiableOn ℝ f (Set.Ioo b c))
+    (h₀ : ∀ x ∈ Set.Ioo a b, deriv f x ≤ 0)
+    (h₁ : ∀ x ∈ Set.Ioo b c, 0 ≤ deriv f x) : IsLocalMin f b := by
+    have Q := @first_derivative_test_Ioo_max (-f) a b c g₀ g₁
+      (by simp_all) (DifferentiableOn.neg hd₀) (DifferentiableOn.neg hd₁)
+      (fun _ => deriv_neg_nonneg hd₀ h₀) (fun _ => deriv_neg_nonpos hd₁ h₁)
+    unfold IsLocalMax IsMaxFilter at Q
+    simp only [Pi.neg_apply, neg_le_neg_iff] at Q
+    exact Q
+
+
+ /-- The First-Derivative Test from calculus, maxima version,
+ expressed in terms of left and right filters. -/
+lemma first_derivative_test_max₀ {f : ℝ → ℝ} {b : ℝ} (h : ContinuousAt f b)
+    (hd₀ : ∀ᶠ x in 𝓝[<] b, DifferentiableAt ℝ f x) (hd₁ : ∀ᶠ x in 𝓝[>] b, DifferentiableAt ℝ f x)
+    (h₀  : ∀ᶠ x in 𝓝[<] b, 0 ≤ deriv f x) (h₁  : ∀ᶠ x in 𝓝[>] b, deriv f x ≤ 0) :
+    IsLocalMax f b := by
+  unfold Filter.Eventually at h₀ h₁ hd₀ hd₁
+  rw [mem_nhdsWithin_Iio_iff_exists_Ioo_subset] at h₀ hd₀
+  rw [mem_nhdsWithin_Ioi_iff_exists_Ioo_subset] at h₁ hd₁
+  obtain ⟨u₀,hu₀⟩ := hd₀; obtain ⟨u₁,hu₁⟩ := hd₁
+  obtain ⟨v₀,hv₀⟩ := h₀; obtain ⟨v₁,hv₁⟩ := h₁
+  apply isLocalMax_of_mono_anti
+  · show max u₀ v₀ < b; exact max_lt (by simp_all) (by simp_all)
+  · show b < min u₁ v₁; exact lt_min (by simp_all) (by simp_all)
+  · exact monotoneOn_of_deriv_nonneg
+      (convex_Ioc _ _)
+      (fun x _ => ContinuousAt.continuousWithinAt ((em (x = b)).elim (fun H => H ▸ h)
+        (fun H => DifferentiableAt.continuousAt (hu₀.2 (by contrapose H;simp_all;linarith)))))
+      (fun x _ => DifferentiableAt.differentiableWithinAt (hu₀.2 (by simp_all)))
+      (fun x _ => by apply hv₀.2;simp_all)
+  · exact antitoneOn_of_deriv_nonpos
+      (convex_Ico _ _)
+      (fun x _ => ContinuousAt.continuousWithinAt ((em (x = b)).elim (fun H => H ▸ h)
+        (fun H => DifferentiableAt.continuousAt (hu₁.2 (by contrapose H;simp_all;linarith)))))
+      (fun x _ => DifferentiableAt.differentiableWithinAt (hu₁.2 (by simp_all)))
+      (fun x _ => by apply hv₁.2;simp_all)
+
+/-- If a set `P` contains left and right neighborhoods of a point `x`
+then `P` contains a punctured neighborhood. -/
+lemma nhdsWithin_punctured_of_Iio_Ioi (P : Set ℝ)
+    (hl : P ∈ nhdsWithin 0 (Set.Iio 0)) (hr : P ∈ nhdsWithin 0 (Set.Ioi 0)) :
+    P ∈ nhdsWithin 0 {0}ᶜ := by
```

Comment:
Why are you suddenly not using notations for `nhdsWithin`? And why are those restricted to neighborhoods of zero? And this has nothing to do in this file. 

Please put
```lean
/-- If a set `P` contains left and right neighborhoods of a point `x` in a linearly ordered
topological space then `P` contains a punctured neighborhood. -/
lemma nhdsWithin_punctured_of_Iio_Ioi {X : Type*} [LinearOrder X] [TopologicalSpace X] 
    {P : Set X} {x : X} (hl : P ∈ 𝓝[<] x) (hr : P ∈ 𝓝[>] x) : P ∈ 𝓝[≠] x := by
  rw [← Iio_union_Ioi, nhdsWithin_union]
  exact Filter.mem_sup.mpr ⟨hl, hr⟩
```
in some relevant file (not the first derivate test file) after checking it does not exist. This lemma is not even used in your PR!

Reply:
I would try to put it next to `nhdsWithin_union` if this works.

Reply:
OK. By the way, regarding [this comment](https://github.com/leanprover-community/mathlib4/pull/16802#discussion_r1789046141), yes I did write a proof that `deriv (-f) = - deriv f` yesterday which I can also add.

Reply:
OK I put it there. It's then only use of `LinearOrder` in that file `ContinuousOn.lean`.